executor/admin.go

```diff
@@ -457,3 +459,248 @@ func (e *RecoverIndexExec) NextChunk(ctx context.Context, chk *chunk.Chunk) erro
 	e.done = true
 	return nil
 }
+
+// CleanupIndexExec represents a cleanup index executor.
+// It is built from "admin cleanup index" statement, is used to delete
+// dangling index data.
+type CleanupIndexExec struct {
+	baseExecutor
+
+	done      bool
+	removeCnt uint64
+
+	index table.Index
+	table table.Table
+
+	idxCols          []*model.ColumnInfo
+	idxColFieldTypes []*types.FieldType
+	idxChunk         *chunk.Chunk
+
+	matchedIndex map[int64]idxData
+	batchSize    uint64
+	batchKeys    []kv.Key
+	idxValsBufs  [][]types.Datum
+	lastIdxKey   []byte
+	scanRowCnt   uint64
+}
+
+type idxData struct {
+	matched bool
+	idxVals []types.Datum
+}
+
+func (e *CleanupIndexExec) getIdxColTypes() []*types.FieldType {
+	if e.idxColFieldTypes != nil {
+		return e.idxColFieldTypes
+	}
+	e.idxColFieldTypes = make([]*types.FieldType, 0, len(e.idxCols))
+	for _, col := range e.idxCols {
+		e.idxColFieldTypes = append(e.idxColFieldTypes, &col.FieldType)
+	}
+	return e.idxColFieldTypes
+}
+
+func (e *CleanupIndexExec) batchGetRecord(txn kv.Transaction) (map[string][]byte, error) {
+	for handle := range e.matchedIndex {
+		e.batchKeys = append(e.batchKeys, e.table.RecordKey(handle))
+	}
+	values, err := txn.GetSnapshot().BatchGet(e.batchKeys)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return values, nil
+}
+
+func (e *CleanupIndexExec) deleteDanglingIdx(txn kv.Transaction, values map[string][]byte) error {
+	for _, k := range e.batchKeys {
+		if _, found := values[string(k)]; !found {
+			_, handle, err := tablecodec.DecodeRecordKey(k)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err := e.index.Delete(e.ctx.GetSessionVars().StmtCtx, txn, e.matchedIndex[handle].idxVals,
+				handle); err != nil {
+				return errors.Trace(err)
+			}
+			e.removeCnt++
+			if e.removeCnt%e.batchSize == 0 {
+				log.Infof("[cleaning up dangling index] table: %v, index: %v, count: %v.",
+					e.table.Meta().Name.String(), e.index.Meta().Name.String(), e.removeCnt)
+			}
+		}
+	}
+	return nil
+}
+
+func (e *CleanupIndexExec) extractIdxVals(row chunk.Row, idxVals []types.Datum) []types.Datum {
```

Comment:
This function logic is the same as RecoverIndexExec's extractIdxVals. Could we use the same one?